esmvaltool/utils/recipe_filler.py

```diff
@@ -0,0 +1,684 @@
+                if 'additional_datasets' in cur_yaml['diagnostics']:
+                    cur_yaml['diagnostics'][diag_var[0]]['variables'][
+                        diag_var[1]]['additional_datasets'].extend(add_dat)
+                else:
+                    cur_yaml['diagnostics'][diag_var[0]]['variables'][
+                        diag_var[1]]['additional_datasets'] = add_dat
+    if cur_yaml:
+        with open(output_recipe, 'w') as yamlfile:
+            yaml.dump(cur_yaml, yamlfile)
+
+
+def _find_all_datasets(recipe_dict, cmip_eras):
+    """Find all datasets explicitly."""
+    datasets = []
+    for cmip_era in cmip_eras:
+        if cmip_era == "CMIP6":
+            activity = "CMIP"
+        else:
+            activity = ""
+        drs, site_path = _get_site_rootpath(cmip_era)
+        if drs in ["default", "SMHI"]:
+            logger.info(f"DRS is {drs}; filter on dataset disabled.")
+            datasets = ["*"]
+        else:
+            if drs in ["BADC", "DKRZ", "CP4CDS"]:
+                institutes_path = os.path.join(site_path, activity)
+            elif drs in ["ETHZ", "RCAST"]:
+                exp = recipe_dict["exp"][0]
+                if exp == "*":
+                    exp = "piControl"  # all institutes have piControl
+                mip = recipe_dict["mip"]
+                var = recipe_dict["short_name"]
+                institutes_path = os.path.join(site_path, exp, mip, var)
+
+            if not os.path.isdir(institutes_path):
+                logger.warning(f"Path to data {institutes_path} "
+                               "does not exist; will look everywhere.")
+                datasets = ["*"]
+                return datasets
+
+            institutes = os.listdir(institutes_path)
+            if drs in ["BADC", "DKRZ", "CP4CDS"]:
+                for institute in institutes:
+                    datasets.extend(
+                        os.listdir(os.path.join(institutes_path, institute)))
+            else:
+                datasets.extend(institutes)
+
+    return datasets
+
+
+def _get_exp(recipe_dict):
+    """Get the correct exp as list of single or multiple exps."""
+    if isinstance(recipe_dict["exp"], list):
+        exps_list = recipe_dict["exp"]
+        logger.info(f"Multiple {exps_list} experiments requested")
+    else:
+        exps_list = [recipe_dict["exp"]]
+        logger.info(f"Single {exps_list} experiment requested")
+
+    return exps_list
+
+
+def _get_datasets(recipe_dict, cmip_eras):
+    """Get the correct datasets as list if needed."""
+    if recipe_dict["dataset"] == "*":
+        datasets = _find_all_datasets(recipe_dict, cmip_eras)
+        return datasets
+    if isinstance(recipe_dict['dataset'], list):
+        datasets = recipe_dict['dataset']
+        logger.info(f"Multiple {datasets} datasets requested")
+    else:
+        datasets = [recipe_dict['dataset']]
+        logger.info(f"Single {datasets} dataset requested")
+
+    return datasets
```

Comment:
```suggestion
    datasets = recipe_dict['dataset']
    if datasets == "*":
        datasets = _find_all_datasets(recipe_dict, cmip_eras)
    elif not isinstance(dataset, list):
        datasets = [datasets]
        logger.info(f"Single {datasets} dataset requested")
    else:
        logger.info(f"Multiple {datasets} datasets requested")

    return datasets
```

Reply:
OK but I like my implementation when it immediately returns `datasets` when that is `*` which is in most of the cases :grin: